In [160]:
import dask.dataframe as dd
import pandas as pd
import numpy as np

1. Create a new dataframe object that consists of samples (i.e., rows) corresponding to states in the US.Create a new dataframe object that consists of samples (i.e., rows) corresponding to states in the US.

In [162]:
data = pd.read_csv("timeseries.csv", dtype={'name':str,'level':str,'city':object,'county':str,'state':str,'country':str,'population':np.float64,'lat':float,'long':float,'url':str,'aggregate':object,'tz':str,'cases':np.float64,'deaths':np.float64,'recovered':np.float64,'active':np.float64,'tested':np.float64,'hospitalized':np.float64,'hospitalized_current':np.float64,'discharged':np.float64,'icu':np.float64,'icu_current':np.float64,'growthFactor':np.float64,'date':object})
df = dd.from_pandas(data)

filtered_df = df.query("country == 'United States'")
filtered_df = filtered_df.groupby("state").count()
filtered_df.compute()

name  level  city  county  country  population  \
state                                                                           
Alabama                        7068   7068     0    6953     7068        7068   
Alaska                         1878   1878     0    1775     1878        1833   
American Samoa                   53     53     0       0       53          53   
Arizona                        1823   1823     0    1661     1823        1823   
Arkansas                       8085   8085     0    7968     8085        8085   
California                     7006   7006     0    6843     7006        7006   
Colorado                       6351   6351     0    6230     6351        6351   
Connecticut                    1021   1021     0     901     1021        1021   
Delaware                        455    455     0     338      455         455   
Florida                        7625   7625     0    7498     7625        7625   
Georgia                       16593  16593     0   16467    16593       16593   
Guam                             54     54     0       0       54          54   
Hawaii                          583    583     0     461      583         583   
Idaho                          3610   3610     0    3495     3610        3610   
Illinois                       9975   9975     0    9811     9975        9975   
Indiana                        9720   9720     0    9598     9720        9720   
Iowa                           9425   9425     0    9306     9425        9425   
Kansas                         8211   8211     0    8090     8211        8211   
Kentucky                      11419  11419     0   11298    11419       11419   
Louisiana                      6915   6915     0    6796     6915        6915   
Maine                          1785   1785     0    1669     1785        1785   
Maryland                       2756   2756     0    2633     2756        2756   
Massachusetts                  1820   1820     0    1664     1820        1803   
Michigan                       8397   8397    40    8279     8397        8357   
Minnesota                      8359   8359     0    8237     8359        8359   
Mississippi                    8678   8678     0    8561     8678        8678   
Missouri                      10360  10360     0   10239    10360       10360   
Montana                        2962   2962     0    2847     2962        2962   
Nebraska                       7630   7630     0    7490     7630        7630   
Nevada                         1518   1518     0    1395     1518        1518   
New Hampshire                  1234   1234     0    1108     1234        1234   
New Jersey                     2547   2547     0    2423     2547        2547   
New Mexico                     3131   3131     0    3014     3131        3131   
New York                       6554   6554     0    6428     6554        6554   
North Carolina                10232  10232     0   10107    10232       10232   
North Dakota                   5133   5133     0    5016     5133        5133   
Northern Mariana Islands         53     53     0       0       53          53   
Ohio                           9199   9199     0    9080     9199        9199   
Oklahoma                       7119   7119     0    6997     7119        7119   
Oregon                         4068   4068     0    3939     4068        4068   
Pennsylvania                   7202   7202     0    7081     7202        7202   
Puerto Rico                      53     53     0       0       53          53   
Rhode Island                   1857   1857  1175    1730     1857         682   
South Carolina                 5063   5063     0    4941     5063        5063   
South Dakota                   4963   4963     0    4845     4963        4963   
Tennessee                      9978   9978     0    9855     9978        9978   
Texas                         22069  22069     0   21924    22069       22069   
United States Virgin Islands     53     53     0       0      

In [163]:
new_column_df = df.query("country == 'United States'")\
    .assign(seen=lambda x: x.recovered + x.active)
new_column_df.compute()

name   level  \
24958                Washoe County, Nevada, United States  county   
24959                Washoe County, Nevada, United States  county   
24960                Washoe County, Nevada, United States  county   
24961                Washoe County, Nevada, United States  county   
24962                Washoe County, Nevada, United States  county   
...                                                   ...     ...   
397823  Detroit City, Wayne County, Michigan, United S...    city   
397824  Detroit City, Wayne County, Michigan, United S...    city   
397825  Detroit City, Wayne County, Michigan, United S...    city   
397826  Detroit City, Wayne County, Michigan, United S...    city   
397827  Detroit City, Wayne County, Michigan, United S...    city   

                city         county     state        country  population  \
24958           <NA>  Washoe County    Nevada  United States    471519.0   
24959           <NA>  Washoe County    Nevada  United States    471519.0   
24960           <NA>  Washoe County    Nevada  United States    471519.0   
24961           <NA>  Washoe County    Nevada  United States    471519.0   
24962           <NA>  Washoe County    Nevada  United States    471519.0   
...              ...            ...       ...            ...         ...   
397823  Detroit City   Wayne County  Michigan  United States         NaN   
397824  Detroit City   Wayne County  Michigan  United States         NaN   
397825  Detroit City   Wayne County  Michigan  United States         NaN   
397826  Detroit City   Wayne County  Michigan  United States         NaN   
397827  Detroit City   Wayne County  Michigan  United States         NaN   

           lat      long                                                url  \
24958   40.582 -119.5885  https://services.arcgis.com/iCGWaR7ZHc5saRIl/a...   
24959   40.582 -119.5885  https://services.arcgis.com/iCGWaR7ZHc5saRIl/a...   
24960   40.582 -119.5885  https://services.arcgis.com/iCGWaR7ZHc5saRIl/a...   
24961   40.582 -119.5885  https://services.arcgis.com/iCGWaR7ZHc5saRIl/a...   
24962   40.582 -119.5885  https://services.arcgis.com/iCGWaR7ZHc5saRIl/a...   
...        ...       ...                                                ...   
397823     NaN       NaN  https://www.michigan.gov/coronavirus/0,9753,7-...   
397824     NaN       NaN  https://www.michigan.gov/coronavirus/0,9753,7-...   
397825     NaN       NaN  https://www.michigan.gov/coronavirus/0,9753,7-...   
397826     NaN       NaN  https://www.michigan.gov/coronavirus/0,9753,7-...   
397827     NaN       NaN  https://www.michigan.gov/coronavirus/0,9753,7-...   

        ... active tested  hospitalized  hospitalized_current  discharged  \
24958   ...    1.0    NaN           NaN                   NaN         NaN   
24959   ...    1.0    NaN           NaN                   NaN         NaN   
24960   ...    NaN    NaN           NaN                   NaN         NaN   
24961   ...    NaN    NaN           NaN                   NaN         NaN   
24962   ...    NaN    NaN           NaN                   NaN         NaN   
...     ...    ...    ...           ...                   ...         ...   
397823  ...    NaN    NaN           NaN                   NaN         NaN   
397824  ...    NaN    NaN           NaN                   NaN         NaN   
397825  ...    NaN    NaN           NaN                   NaN         NaN   
397826  ...    NaN    NaN           NaN                   NaN         NaN   
397827  ...    NaN    NaN           NaN                   NaN         NaN   

        icu  icu_current  growthFactor        date  seen  
24958   NaN          NaN           NaN  2020-03-05   NaN  
24959   NaN          NaN           1.0  2020-03-06   NaN  
24960   NaN          NaN           1.0  2020-03-07   NaN  
24961   NaN          NaN           2.0  2020-03-08   NaN  
24962   NaN          NaN           1.0  2020-03-09   NaN  
...     ...          ...           ...         ...   ...  
397823  NaN          NaN

Interpretation: Here, I filtered the dataset to include only records where the country is the United States using the query function. Since the dataset contains global COVID-19 data, this step ensures that only U.S.-specific data is considered for analysis. Next, it creates a new column called seen, which is the sum of recovered and active cases. This metric helps in understanding the total number of observed cases, excluding deaths, and can be useful in analyzing trends in recovery versus active cases. Since Dask operates in a lazy evaluation mode, the .compute() function is used to execute the computations and return a tangible result as a Pandas DataFrame. 

2. During the time period 2020-Jan-01 to 2021-Feb-28, rank states in terms of their per-capita mortality? Compute per-capita mortality during a specific period as the ratio of total deaths during that time period, to the average population of the state (compute the average population during the time period).

In [204]:
# Convert date column to datetime
new_column_df["date"] = dd.to_datetime(new_column_df["date"])

# Filter for the specified time period
time_period_df = new_column_df[
    (new_column_df["date"] >= "2020-01-01") & (new_column_df["date"] <= "2021-02-28")
]

# Group by state and compute total deaths and average population
state_stats = time_period_df.groupby("state").agg(
    total_deaths=("deaths", "sum"),
    avg_population=("population", "mean")
).compute()

# Compute per-capita mortality
state_stats["per_capita_mortality"] = state_stats["total_deaths"] / state_stats["avg_population"]

# Rank states by per-capita mortality
state_stats["rank"] = state_stats["per_capita_mortality"].rank(ascending=False)
print(state_stats[["per_capita_mortality", "rank"]].sort_values("rank"))

                              per_capita_mortality  rank
state                                                   
New York                                  6.395701   1.0
Michigan                                  3.204753   2.0
Louisiana                                 2.735288   3.0
Illinois                                  2.043863   4.0
New Jersey                                2.031200   5.0
Georgia                                   2.026085   6.0
Pennsylvania                              1.831572   7.0
Virginia                                  1.359063   8.0
Mississippi                               1.356705   9.0
Indiana                                   1.353423  10.0
Ohio                                      1.058943  11.0
Iowa                                      0.958771  12.0
Massachusetts                             0.904775  13.0
Colorado                                  0.894296  14.0
Minnesota                                 0.768853  15.0
Kentucky                       

Here, I used COVID-19 data to compute and rank U.S. states based on per-capita mortality during the period from January 1, 2020, to February 28, 2021. First, I converted the date column into a datetime format to facilitate time-based filtering. The dataset was then restricted to the specified time range, ensuring that only relevant records were analyzed. The code grouped the data by state, calculating the total number of deaths and the average population for each state. Using these values, per-capita mortality was computed as the ratio of total deaths to the average population, providing a normalized measure of mortality risk across states. The states were then ranked in descending order based on per-capita mortality, with lower ranks indicating higher mortality rates. The results indicate that New York experienced the highest per-capita mortality during this period, followed by Michigan, Louisiana, and Illinois.

3. During the same time period, compute the case fatality rate (CFR) per month, using one of the approaches defined in this scientific brief from the World Health Organization: https://www.who.int/news-room/commentaries/detail/estimating-mortality-from-covid-19Links to an external site.. This computation should yield an array of dimensions 50 (states) X 14 (months). State the assumptions you are making in computing this metric.

In [169]:
# Extract year and month from the date column
time_period_df["date"] = dd.to_datetime(time_period_df["date"])
time_period_df["month"] = time_period_df["date"].dt.to_period("M")

# Group by state and month, and compute total deaths and cases
monthly_cfr = time_period_df.groupby(["state", "month"]).agg(
    total_deaths=("deaths", "sum"),
    total_cases=("cases", "sum")
).compute()

# Compute CFR
monthly_cfr["cfr"] = (monthly_cfr["total_deaths"] / monthly_cfr["total_cases"]) * 100

cfr_matrix = monthly_cfr.pivot_table(index="state", columns="month", values="cfr")
cfr_matrix = cfr_matrix.fillna(0)
print("Case Fatality Rate Matrix:")
cfr_matrix

Case Fatality Rate Matrix:


month                         2020-01   2020-02   2020-03    2020-04  \
state                                                                  
Alabama                           0.0  0.000000  0.532313   2.830899   
Alaska                            0.0  0.000000  0.335008   2.314519   
Arizona                           0.0  0.000000  0.000000   1.486545   
Arkansas                          0.0  0.000000  0.915656   1.911450   
California                        0.0  0.000000  2.006735   3.479974   
Colorado                          0.0  0.000000  0.939250   2.636616   
Connecticut                       0.0  0.000000  1.814771   6.477626   
Delaware                          0.0  0.000000  1.334107   2.734038   
Florida                           0.0  0.000000  0.842669   2.905738   
Georgia                           0.0  0.000000  2.185574   3.980977   
Guam                              0.0  0.000000  2.702703   3.615483   
Hawaii                            0.0  0.000000  0.036271   0.000000   
Idaho                             0.0  0.000000  1.448967   2.258980   
Illinois                          0.0  0.000000  1.288155   4.048394   
Indiana                           0.0  0.000000  2.396460   4.775795   
Iowa                              0.0  0.000000  0.863465   2.326472   
Kansas                            0.0  0.000000  0.989866   2.364135   
Kentucky                          0.0  0.000000  2.225112   5.074331   
Louisiana                         0.0  0.000000  3.916843   5.191232   
Maine                             0.0  0.000000  0.529355   3.778067   
Maryland                          0.0  0.000000  0.261008   3.684368   
Massachusetts                     0.0  0.000000  0.499505   4.415282   
Michigan                          0.0  0.000000  2.493512   7.508176   
Minnesota                         0.0  0.000000  0.000000   2.527566   
Mississippi                       0.0  0.000000  1.504930   3.296058   
Missouri                          0.0  0.000000  1.196568   2.878952   
Montana                           0.0  0.000000  0.000000   0.966368   
Nebraska                          0.0  0.000000  0.000000   1.422412   
Nevada                            0.0  0.000000  2.033319   4.051924   
New Hampshire                     0.0  0.000000  0.268757   1.546978   
New Jersey                        0.0  0.000000  0.401749   4.064989   
New Mexico                        0.0  0.000000  0.705290   2.949834   
New York                          0.0  0.000000  2.075213   6.525716   
North Carolina                    0.0  0.000000  0.331498   2.722044   
North Dakota                      0.0  0.000000  0.253646   0.893946   
Northern Mariana Islands          0.0  0.000000  0.000000  14.468085   
Ohio                              0.0  0.000000  1.529431   4.404734   
Oklahoma                          0.0  0.000000  2.833876   5.439956   
Oregon                            0.0  0.000000  2.660322   3.617323   
Pennsylvania                      0.0  0.000000  1.081428   3.292275   
Puerto Rico                       0.0  0.000000  0.000000   5.808946   
Rhode Island                      0.0  0.000000  0.061703   1.516496   
South Carolina                    0.0  0.000000  1.964286   2.882223   
South Dakota                      0.0  0.000000  0.775194   0.453175   
Tennessee                         0.0  0.000000  0.036523   1.964376   
Texas                             0.0  0.000000  1.184979   2.403974   
United States Virgin Islands      0.0  0.000000  0.000000   5.223881   
Utah                              0.0  0.000000  0.179292   0.446306   
Vermont                           0.0  0.000000  3.411633   4.238581   
Virginia                          0.0  0.000000  1.406402   2.865308   
Washington                        0.0  1.360544  5.051079   5.255217   
Washington, D.C.                  0.0  0.000000  1.436880   3.736434   
West Virginia                     0.0  0.000000  0.000000   2.181204   
Wisconsin                         0.0  0

In computing the Case Fatality Rate (CFR) per month for each U.S. state, I have assumed that the likelihood of detecting COVID-19 cases and deaths remains consistent over the outbreak period. Early in the pandemic, surveillance systems prioritized symptomatic individuals seeking care, leading to a potential overestimation of CFR as asymptomatic and mild cases were underreported. As testing capacity and case definitions improved, case detection likely became more comprehensive, reducing this bias over time. To mitigate early-stage overestimation, one potential approach is left censoring removing cases from the analysis that occurred before the establishment of robust surveillance systems. The generated CFR matrix, structured with states as rows and months as columns, allows for trend analysis of case fatality rates across different regions and timeframes, helping to identify shifts in mortality risk and healthcare response effectiveness.

4. Using this matrix as input, compute the ranking of states, on how the CFR rate changed over time. This computation will involve an aggregation of of month-to-month changes in CFR. Note that some of these month-to-month changes can be positive (CFR increases from a previous month to the current month) or negative (CFR decrease from a previous month to the current month), or zero (no change in CFR). You need to aggregate these individual changes, across all time periods.

In [172]:
# Compute month-to-month changes in CFR
cfr_matrix_diff = cfr_matrix.diff(axis=1)  
cfr_change_aggregated = cfr_matrix_diff.sum(axis=1)
cfr_change_aggregated = cfr_change_aggregated.sort_values(ascending=False)
print("cfr_change_aggregated:")
cfr_change_aggregated

cfr_change_aggregated:


state
Connecticut                     9.500807
Michigan                        8.623360
New Jersey                      8.251406
Massachusetts                   7.437951
Pennsylvania                    7.418973
New York                        7.233961
Northern Mariana Islands        6.535948
New Hampshire                   6.436738
United States Virgin Islands    5.976096
Rhode Island                    5.821287
Washington, D.C.                5.330110
Indiana                         5.314359
Ohio                            5.275945
Louisiana                       5.138853
Colorado                        5.063691
Illinois                        4.802791
Maryland                        4.649307
Vermont                         4.559147
Delaware                        4.013148
New Mexico                      4.003795
Missouri                        3.979086
Minnesota                       3.958075
Washington                      3.921734
Mississippi                     3.743453
Kentucky  

Interpretation: The month-to-month changes in Case Fatality Rate (CFR) reveal significant variations across states, with some experiencing substantial increases while others remained relatively stable. Connecticut, Michigan, and New Jersey saw the highest CFR changes, exceeding 8.0, followed closely by Massachusetts, Pennsylvania, and New York. The Northern Mariana Islands and U.S. Virgin Islands also exhibited notable fluctuations, suggesting considerable month-to-month variability. States such as Indiana, Ohio, Louisiana, and Colorado had moderate CFR changes, ranging between 5.0 and 6.0, while others like Florida, North Carolina, and Kansas saw smaller increases, between 1.5 and 2.5. Interestingly, Hawaii reported no change in CFR, indicating complete stability during the analyzed period.

4. For each of the above operations, explain using appropriate reasoning, whether using a parallelized and/or distributed way of performing the computation makes sense.

Using a parallelized and/or distributed approach for computing the CFR trends across states is justified based on the operations performed. Since the dataset involves COVID-19 case and death counts across multiple states and months, parallelization helps efficiently execute computations that can be performed independently, such as grouping by state and month, summing total deaths and cases, and computing per-capita mortality. The use of Dask here inherently supports parallelization, making operations like .compute() well-suited for handling large datasets. Additionally, when aggregating month-to-month CFR changes for ranking states, parallelization can speed up the process since each state's computations do not depend on others. Distributed computing becomes relevant when dealing with datasets too large to fit into memory, as it enables the workload to be spread across multiple computing nodes. However, if the dataset size is relatively small and can be processed efficiently in memory, the overhead of distributed computing might not be necessary. Given that this requires computing CFR across all states over a 7-month period, leveraging parallelism in Dask is an optimal choice to improve efficiency while maintaining scalability.